src/future/batch_semaphore.rs

```diff
@@ -447,6 +492,15 @@ impl BatchSemaphore {
 
         state.waiters.remove(index).unwrap();
         assert!(waiter.is_queued.swap(false, Ordering::SeqCst));
+
+        if matches!(self.fairness, Fairness::StrictlyFair) && index == 0 {
+            // If the semaphore is strictly fair, and we removed the first waiter, check if its
+            // removal unblocks remaining waiters.  This can happen in the following situation:
+            // - the semahore has 2 permits available
+            // - there are 3 waiters W1, W2, W3 where W1 wants 3 permits, and W2,W3 want 1 permit each
+            // - if W3 is removed (because it drops out), we want to ensure W2,W3 are granted the semaphore
```

Comment:
Nit: W2, W3 or W2 and W3

Nit: "ensure that"

Reply:
Nit: "if W3 is removed" -> "if W1 is removed"

Reply:
Fixed